[READY][API] Return debugging information as a dictionary instead of a string

micbou opened the pull request:
Currently, the data returned by the `debug_info` handler is a block of text. This makes it difficult for clients to retrieve a specific information on the server (e.g. the path of the logfiles, the process identifier of a sub-server, etc.) or to customize the way it is displayed to the user.

We improve this by returning a dictionary instead of a string. Its structure is fully explained in [the new `debug_info` API documentation](http://micbou.github.io/ycmd/#path--debug_info).

There are two important changes on its content:
- add information on the Python interpreter used to start the ycmd server: its path and its version.
- move information on the `.ycm_extra_conf.py` file (its path and if it is loaded or not) from the Clang completer to ycmd itself since it is not supposed to be specific to this completer.

This is obviously an API-breaking change but not a big one since it only affects the `debug_info` handler.

@abingham @Qusic @LuckyGeck @mawww @richard1122

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/601)

<!-- Reviewable:end -->

vheon commented:
Review status: 0 of 21 files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/tests/clang/debug_info_test.py, line 67 at r1](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnwkUocj95LwMAhab5:-KRnwkUocj95LwMAhab6:b-xhgulj) ([raw file](https://github.com/valloric/ycmd/blob/909175ebe2ce2610337eb0af285d4b69bd423024/ycmd/tests/clang/debug_info_test.py#L67)):*
> ```Python
>     } ) )
>   )
>   # Second time, an empty list of flags is returned.
> ```

isn't this the same response that we have above? I think that since you moved the extra_conf stuff to be always returned we should at least rename the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: 0 of 21 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnzYmJsSlxf7w-DFF8:-KRnzYmKXjNyZmKEq6PX:b8r0kep) (no related file):*
I understand that now we return all the information all the time since debug info is now a structure but before, at least in the typescript, rust, python, javascript tests, we were actually testing that also the values had a meaning: for example here we were testing that the string reporting the status of the server were actually correct (running/not running) while now we check only the structure of the `debug_info` handle. Is this something worth keeping?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 20 of 21 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnzYmJsSlxf7w-DFF8:-KRoRxIJWEm9pXm_cGol:bdggsdb) (no related file):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I understand that now we return all the information all the time since debug info is now a structure but before, at least in the typescript, rust, python, javascript tests, we were actually testing that also the values had a meaning: for example here we were testing that the string reporting the status of the server were actually correct (running/not running) while now we check only the structure of the `debug_info` handle. Is this something worth keeping?
</details>
I think it is still useful to check if the `debug_info[ 'completer' ]` structure is correct. Also, without those tests, the `DebugInfo` method in the completers would not be covered.

---

*[ycmd/tests/clang/debug_info_test.py, line 67 at r1](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnwkUocj95LwMAhab5:-KRoTh8PmYMUGLR2ry_W:b-4j0r14) ([raw file](https://github.com/valloric/ycmd/blob/909175ebe2ce2610337eb0af285d4b69bd423024/ycmd/tests/clang/debug_info_test.py#L67)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> isn't this the same response that we have above? I think that since you moved the extra_conf stuff to be always returned we should at least rename the test.
</details>
Yes, this is the same response but the executed code is different. When sending the first request, the `FlagsForFile` function raises a `NoExtraConfDetected` exception that we catch in `DebugInfo`. Second request, the `FlagsForFile` function returns `None`. We handle both cases by setting the flags to an empty list.

I updated and renamed the tests to make it clear that we are checking the flags.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnzYmJsSlxf7w-DFF8:-KRod2IOPqESK3zWPOjK:bvdkhta) (no related file):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> I think it is still useful to check if the `debug_info[ 'completer' ]` structure is correct. Also, without those tests, the `DebugInfo` method in the completers would not be covered.
</details>
I'm not saying that we should not check the structure, I'm saying that I'm feeling that we're loosing some coverage because we're not actually checking behaviour anymore because all we check is that the structure is the right one and that's it. But I think that is just a feeling because there is no actual behaviour to test anymore since we have removed all the `if` statement in the `DebugInfo` method :confused: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnzYmJsSlxf7w-DFF8:-KRof7tjRFd4AVvP22li:b-weq3sp) (no related file):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I'm not saying that we should not check the structure, I'm saying that I'm feeling that we're loosing some coverage because we're not actually checking behaviour anymore because all we check is that the structure is the right one and that's it. But I think that is just a feeling because there is no actual behaviour to test anymore since we have removed all the `if` statement in the `DebugInfo` method :confused: 
</details>
Yes, this is why I removed the tests. Now that there is only one branch in `DebugInfo` (except for the Clang and C# completers, which have more tests), we only need one test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnzYmJsSlxf7w-DFF8:-KRooI6JDGw5hqkI-uAE:b-n180dh) (no related file):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Yes, this is why I removed the tests. Now that there is only one branch in `DebugInfo` (except for the Clang and C# completers, which have more tests), we only need one test.
</details>
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Looks nice!

I keep seeing these draft API docs; they look really sweet! When can we see some PRs for that? :) 

---

*[ycmd/handlers.py, line 234 at r2](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRoo30jKZuRAm8l-MTL:-KRoo30jKZuRAm8l-MTM:bmb54fe) ([raw file](https://github.com/valloric/ycmd/blob/fd756a380330166e7601b5a84fc46b1f5ab9a14c/ycmd/handlers.py#L234)):*
> ```Python
>     'extra_conf': {
>       'path': extra_conf_path,
>       'loaded': loaded
> ```

I'd rather make this something like `is_loaded`. More obvious it's a bool.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/handlers.py, line 234 at r2](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRoo30jKZuRAm8l-MTL:-KRopYuCi3ehQdsL5Bov:b-njqt2) ([raw file](https://github.com/valloric/ycmd/blob/fd756a380330166e7601b5a84fc46b1f5ab9a14c/ycmd/handlers.py#L234)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> I'd rather make this something like `is_loaded`. More obvious it's a bool.
</details>
I'll change it. Should we replace `support` by `has_support` in Clang info?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 19 of 21 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/handlers.py, line 234 at r2](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRoo30jKZuRAm8l-MTL:-KRtcOSC7HrpuAxWP1-d:b-llc7ev) ([raw file](https://github.com/valloric/ycmd/blob/fd756a380330166e7601b5a84fc46b1f5ab9a14c/ycmd/handlers.py#L234)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> I'll change it. Should we replace `support` by `has_support` in Clang info?
</details>
Yep, that sounds like a good change too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/handlers.py, line 234 at r2](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRoo30jKZuRAm8l-MTL:-KRtnlKETjzK8TytOK7-:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/fd756a380330166e7601b5a84fc46b1f5ab9a14c/ycmd/handlers.py#L234)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Yep, that sounds like a good change too.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #629) made this pull request unmergeable. Please resolve the merge conflicts.

vheon commented:
@micbou What is blocking this PR?

micbou commented:
@vheon Nothing. I am going to fix the conflicts then I'll mark it as ready.

micbou commented:
Updated with minor changes:
 - use [`platform.python_version()`](https://docs.python.org/2/library/platform.html#platform.python_version) instead of [`sys.version`](https://docs.python.org/2/library/sys.html#sys.version) to get the Python version; `sys.version` returns additional information that's not really useful (date of compilation and compiler used);
 - rename `description` to `key` in `DebugInfoItem` object.

[API documentation](http://micbou.github.io/ycmd/#path--debug_info) updated as well.

PR is now ready.

---

Reviewed 8 of 8 files at r4, 12 of 12 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KVGZay3RcRpfQdXYzO-:b-idnw4l)*
<!-- Sent from Reviewable.io -->

Valloric commented:
> platform.python_version()

I think we used to use that at some point but moved away from it. Can't really remember why. 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KVGofvOg5TZc4jVYKvg:b-or7u3i)*
<!-- Sent from Reviewable.io -->

micbou commented:
I looked at [the Python sources](https://github.com/python/cpython/blob/master/Lib/platform.py#L1234) and this function simply parses the string returned by `sys.version` so it should be completely safe.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KVGrSAqq6NAVV-IZAGz:b-vyb7es)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@micbou I think it might be related to the Python interpreter embedded inside Vim. Especially for the Windows Vim version.

Anyway, if you're confident this is solid, I'm fine with it. Can't really provide an argument based on vague memory. :)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KVIHZj9viJuSBKQPQaT:b-txz30q)*
<!-- Sent from Reviewable.io -->

micbou commented:
If this is related to Vim then we shouldn't worry since we are doing that in the ycmd server.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KVK_ic_7sJs1nkkVO-n:b-vm11nw)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #635) made this pull request unmergeable. Please resolve the merge conflicts.

micbou commented:
I've fixed the conflicts. I think we can merge this PR if nobody is against it. This new API is really convenient for adding more debugging information in completers which I am planning to do for the Python completer (it will be most likely information on the project root and the preloaded modules). In addition, this makes information about the `.ycm_extra_conf.py` file not tied anymore to the Clang completer but to ycmd itself so I will not have to add this information in the Python completer.

---

Reviewed 3 of 3 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KX1Lc3vkzcnD1sbj1MT:b-chr3du)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Still a fan of this PR. :)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KX8bw7cR9vw_QCIqocV:b3jka6d)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #680) made this pull request unmergeable. Please resolve the merge conflicts.

micbou commented:
Fixed the conflicts and updated the copyrights (new year).

---

Reviewed 2 of 2 files at r8, 21 of 21 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KaSd9zd9u4NWUbjOK-7:b-l2cfke)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 21 of 21 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601#-:-KaSqhE5IIE0TXwXXvEu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/ycmd/601#-KRnzYmJsSlxf7w-DFF8:-KaiJC-lzRRWIdANMCim:b-7ennmq) (no related file):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

:lgtm: Looks nice!

I keep seeing these draft API docs; they look really sweet! When can we see some PRs for that? :) 
</blockquote></details>

Let's go! :)

@homu r+

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/601)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit bc258dd has been approved by `Valloric`

<!-- @homu r=Valloric bc258ddc86ab20c32310fcadf84dc2f3241d1c02 -->

homu commented:
:hourglass: Testing commit bc258dd with merge ec4f335...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/192871502)
